**The detailed documentation of Polaris and its usage can be found at this link:** https://nucleome-polaris.readthedocs.io/en/latest

**The Hugging Face page of Polaris is:** https://huggingface.co/rr-ss/Polaris , where the **weights file** and **exapmle mcool** to run Polaris can be found.

**The tutorials can be found under the example folder**.

<img src="./doc/logo.png" alt="Polaris" title="Polaris" width="400">

# A Universal Framework for Chromatin Loop Annotation from Bulk and Single-cell Contact Maps

<div style="text-align: center;">
    <a href="https://github.com/ai4nucleome/Polaris">README (EN)</a>  &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;     
    <a href="https://github.com/ai4nucleome/Polaris/tree/master/doc">README (CN)</a>
</div>


&nbsp;&nbsp;&nbsp; 

<div style="text-align: center;">
    <a href="https://github.com/ai4nucleome/Polaris/releases/latest">
   <img src="https://img.shields.io/badge/Polaris-v1.1.0-green">
   <img src="https://img.shields.io/badge/platform-Linux%20%7C%20Mac%20-green">
   <img src="https://img.shields.io/badge/Language-python3-green">
   <!-- <img src="https://img.shields.io/badge/dependencies-tested-green"> -->
</a>  
</div>


üåü **Polaris** is a versatile and efficient command line tool tailored for rapid and accurate chromatin loop detectionfrom from contact maps generated by various assays, including bulk Hi-C, scHi-C, Micro-C, and DNA SPRITE. Polaris is particularly well-suited for analyzing **sparse scHi-C data and low-coverage datasets**.

<div style="text-align: center;">
    <img src="./doc/Polaris.png" alt="Polaris Model" title="Polaris Model" width="600">
</div>


- Using examples for single cell Hi-C and bulk cell Hi-C loop annotations are under [**example folder**](https://github.com/ai4nucleome/Polaris/tree/master/example).
- The scripts and data to **reproduce our analysis** can be found at: [**Polaris Reproducibility**](https://zenodo.org/records/14294273).

> ‚ùóÔ∏è<b>NOTE‚ùóÔ∏è:</b> We suggest users run Polaris on <b>GPU</b>. 
> You can run Polaris on CPU for loop annotations, but it is much slower than on GPU. 
> If you encounter a `CUDA OUT OF MEMORY` error, please: 
> - Check your GPU's status and available memory.
> - Reduce the --batchsize parameter. (The default value of 128 requires approximately 36GB of CUDA memory. Setting it to 24 will reduce the requirement to less than 10GB.)

## [üìùDocumentation](https://nucleome-polaris.readthedocs.io/en/latest/)
 **Detailed documentation** can be found at: [Polaris Doc](https://nucleome-polaris.readthedocs.io/en/latest/).

## Installation
Polaris is developed and tested on Linux machines with python3.9 and relies on several libraries including pytorch, scipy, etc. 
We **strongly recommend** that you install Polaris in a virtual environment.

We suggest users using [conda](https://anaconda.org/) to create a virtual environment for it (It should also work without using conda, i.e. with pip). You can run the command snippets below to install Polaris:

```bash
git clone https://github.com/ai4nucleome/Polaris.git
cd Polaris
conda create -n polaris python=3.9
conda activate polaris
```
-------
### Package dependencies
Polaris relies on the following packages:
```setuptools==75.1.0
appdirs==1.4.4
click==8.0.1
cooler==0.8.11
matplotlib==3.8.0
numpy==1.22.4
pandas==1.3.0
scikit-learn==1.4.2
scipy==1.7.3
timm==0.6.12
tqdm==4.65.0
```
Please install PyTorch == 2.2.2 according to its official [documentation](https://pytorch.org/get-started/previous-versions/). We recommend using [PyTorch 2.2.2](https://pytorch.org/get-started/previous-versions/#linux-and-windows-23) for best compatibility.


### Install Polaris:
```bash
./setup.sh
```
It will automatically download Polaris model's weights from [Hugging Face](https://huggingface.co/rr-ss/Polaris) and install Polaris.

You can also download model's weights file manually from [there](https://huggingface.co/rr-ss/Polaris/resolve/main/polaris/model/sft_loop.pt?download=true) and put it in `Polaris/polaris/model` and change the file name to `sft_loop.pt`.



The installation requires network access to download libraries. Usually, the installation will finish within 3 minutes. The installation time is longer if network access is slow and/or unstable.

## Quick Start for Loop Annotation
**Detailed documentation** can be found at üõú **this link**: [Polaris Doc](https://nucleome-polaris.readthedocs.io/en/latest/) üõú.

**For detailed documentation or parameter setting, please run:**
```bash
polaris --help
```
or **check the instruction [here](https://nucleome-polaris.readthedocs.io/en/latest/)**.

---
To quick run **Polaris** at **5kb** resolution with **default parameters**, you can use the command snippets below:
```bash
polaris loop pred -i [input mcool file] -o [output path of annotated loops]
```
`-i` indicates the path of input mcool file; `-o` indicates the path of output path of detected loops in bed format; At default parameters, polaris will detected loops from input contact map at **5kb** resolution for all autosomes.

The more **detailed parameter instructions** can be found at this link: üõú [Polaris Doc at ReadTheBook](https://nucleome-polaris.readthedocs.io/en/latest/CLI_reference.html#) üõú

---
### output format
It contains tab separated fields as follows:
```
Chr1    Start1    End1    Chr2    Start2    End2    Score
```
|     Field     |                                  Detail                                 |
|:-------------:|:-----------------------------------------------------------------------:|
|   Chr1/Chr2   | chromosome names                                                        |
| Start1/Start2 | start genomic coordinates                                               |
|   End1/End2   | end genomic coordinates (i.e. End1=Start1+resol)                        |
|     Score     | Polaris's loop score [0~1]                                              | 


## Citation:
Yusen Hou, Audrey Baguette, Mathieu Blanchette*, & Yanlin Zhang*. __A versatile tool for chromatin loop annotation in bulk and single-cell Hi-C data__. _bioRxiv_, 2024. [Paper](https://doi.org/10.1101/2024.12.24.630215)
<br>
```
@article {Hou2024Polaris,
	title = {A versatile tool for chromatin loop annotation in bulk and single-cell Hi-C data},
	author = {Yusen Hou, Audrey Baguette, Mathieu Blanchette, and Yanlin Zhang},
	journal = {bioRxiv}
	year = {2024},
}
```

## üì© Contact
A GitHub issue is preferable for all problems related to using Polaris. 

For other concerns, please email Yusen Hou or Yanlin Zhang (yhou925@connect.hkust-gz.edu.cn,  yanlinzhang@hkust-gz.edu.cn).